core/src/main/scala/chisel3/Module.scala

```diff
@@ -394,17 +394,34 @@ package experimental {
     private[chisel3] def isClosed = _closed
 
     /** Mutable state that indicates if IO is allowed to be created for this module.
+      *  - List.empty: IO creation is allowed
+      *  - List.nonEmpty: IO creation is not allowed at contained location(s)
       * This can be used for advanced Chisel library APIs that want to limit
       * what IO is allowed to be created for a module.
       */
-    private var _isIOCreationAllowed = true
+    private[chisel3] var _whereIOCreationIsDisallowed: List[SourceInfo] = Nil
 
     /** If true, then this module is allowed to have user-created IO. */
-    private[chisel3] def isIOCreationAllowed = _isIOCreationAllowed
+    private[chisel3] def isIOCreationAllowed = _whereIOCreationIsDisallowed.isEmpty
 
     /** Disallow any more IO creation for this module. */
-    private[chisel3] def disallowIOCreation(): Unit = {
-      _isIOCreationAllowed = false
+    def disallowIOCreation()(implicit si: SourceInfo): Unit = {
+      _whereIOCreationIsDisallowed = si +: _whereIOCreationIsDisallowed
+    }
```

Comment:
This public API is duplicated by `endIOCreation`. Can only one of them be public to avoid user confusion?

Reply:
Thanks for adding `SourceInfo` to this!